Adding userpool and updating documentation (#409)
tnc-ca-geo/animl-api: documentation/NewStack.md

# Deploying a New Animl Stack

For a comprehensive overview of the Animl architecture and how its services are integrated, see the [architecture documentation](README.md).

> 

Reviewer: We should mention that we currently only support Amazon Web Services (AWS) deployments. Which IMO i snot self-evident.

> In order to create a new instance of the entire Animl stack, we need to deploy things in a certain order to make sure

Reviewer: Specify the 
"things" in a step list and structure the rest accordingly (numbered steps).

everything is working properly.

## Prerequisites

- An AWS account with admin permissions
- [Node.js & npm](https://nodejs.org/)
- [Serverless Framework](https://www.serverless.com/framework/docs/getting-started/)
- [AWS CLI](https://aws.amazon.com/cli/) configured with an `animl` profile
- [Docker](https://docs.docker.com/engine/install/)

> - A [MongoDB Atlas](https://www.mongodb.com/cloud/atlas) subscription (via AWS Marketplace)

Reviewer: Not exactly a pre-requisite since we are describing how to create it.

> - Access to the following repositories:

Reviewer: ```suggestion
- Access to the following repositories (all of them are public repositories):
```

  - [animl-api](https://github.com/tnc-ca-geo/animl-api)
  - [animl-ingest](https://github.com/tnc-ca-geo/animl-ingest)
  - [animl-frontend](https://github.com/tnc-ca-geo/animl-frontend)
  - [exif-api](https://github.com/tnc-ca-geo/exif-api)
  - [animl-ml](https://github.com/tnc-ca-geo/animl-ml) (for ML model deployment)

> ## Deployment Steps

Reviewer: I think we are currently bound to particular regions (e.g. us-west-2) because all the services need to be in one region, but they are not provided in every region (Cognito in particular). We basically need to either suggest us-west-2, find out what other regions would be work, or create the interfaces such that certain services could live in different regions (which they in priniciple can). But people might fail if the start the process in the wrong region.

> 1. A MongoDB cluster instance must exist to manage all the resources needed to run this

Reviewer: ```suggestion
1. The project requires a MongoDB cluster. If you need to create one see MongoDB.md (link). Before you continue you need to create an SSM parameter in AWS System Managers parameter store (https://us-west-1.console.aws.amazon.com/systems-manager/parameters/) holding the connection string (URL). The key needs to be /db/mongo-db-url-{deployment-stage} and the connection string should have the form ... (best provide also an complete example.
```

Clarify further what [env] in your example means. A user-selectable name for the deployment stage like dev/stage/prod or the cloudformation stack name.

> API. Once this has been created, a SSM parameter named `/db/mongo-db-url-[env]` must

Reviewer: Explain where the [env] value comes from and how it is set.

> be created with the MongoDB connection string as its value. To learn more about of how

Reviewer: Provide example for MongoDB connection string and specify, e.g.

The connections string is an URL that will point the application to the database.

to setup your MongoDB instance reference the documentation [here](Mongo.md).

> After setting up your Mongo instance, you will need to seed your DB which can be found [here](../README.md#seeding-db)

Reviewer: Are we able to do that right away or does that require other AWS services in place?

Author: Yes immediately after the MongoDB setup, you can test the seeding script.

> 2. We currently depend on this CloudFormation Template Stack that is managed by

Reviewer: ```suggestion
2. In the next step, we need to create a User pool in AWS Cognito. The whole setup is managed by an AWS  Cloudformation template ```Userpool.yml```

Reviewer: UserPool.yml Can we avoid capitalization in filenames outside of specific programming language conventions such as JS or C. Or at least stick to real camelCase.

[UserPool.yml](../UserPool.yml) that creates and manages all of the resources related
to Auth. To deploy this stack you need to run this command with the proper permissions enabled:

    ```
    aws cloudformation deploy --template-file UserPool.yml  --stack-name animl-user-pool --parameter-overrides Name=animl-dev UsePreauth=false --capabilities CAPABILITY_NAMED_IAM
    ```

    This deployment will also add the necessary SSM parameter that the API will reference.
    Be sure to create versions for all envs you plan on deploying.

> 3. Now we can deploy animl-api by following the instructions [here](../README.md#local-testing-and-dev-deployment).

Reviewer: Can you please specify what we are installing. E.g.

At this point, we are ready to install the Animl API (https://github.com/tnc-ca-geo/animl-api {full Github link}).
The Animl API is the centerpiece of the application it stores incoming data, triggers image inference, and serves the data to the frontend.

>     But when deploying a new animl-api isntance, you will have to comment out all SSM params prefixed by `ml/` in the

Reviewer: Is there a way to fix this workaround? @nathanielrindlaub?
